File: design_patterns_101/Behavioral Patterns/Template Method Pattern/templateMethodPattern.js
// Abstract Class
class AbstractClass {
    templateMethod() {
      this.baseOperation1();
      this.requiredOperation1();
      this.baseOperation2();
      this.hook1();
      this.requiredOperation2();
      this.baseOperation3();
      this.hook2();
    }
  
    baseOperation1() {
      console.log("AbstractClass says: I am doing the bulk of the work");
    }
  
    baseOperation2() {
      console.log("AbstractClass says: But I let subclasses override some operations");
    }
  
    baseOperation3() {
      console.log("AbstractClass says: But I am doing the bulk of the work");
    }
  
    requiredOperation1() {
      throw new Error("Abstract method!");
    }
  
    requiredOperation2() {
      throw new Error("Abstract method!");
    }
  
    hook1() {}
  
    hook2() {}
  }
  
  // Concrete Class
  class ConcreteClass1 extends AbstractClass {
    requiredOperation1() {
      console.log("ConcreteClass1 says: Implemented Operation1");
    }
  
    requiredOperation2() {
      console.log("ConcreteClass1 says: Implemented Operation2");
    }
  }
  
  // Another Concrete Class
  class ConcreteClass2 extends AbstractClass {
    requiredOperation1() {
      console.log("ConcreteClass2 says: Implemented Operation1");
    }
  
    requiredOperation2() {
      console.log("ConcreteClass2 says: Implemented Operation2");
    }
  
    hook1() {
      console.log("ConcreteClass2 says: Override Hook1");
    }
  
    hook2() {
      console.log("ConcreteClass2 says: Override Hook2");
    }
  }
  
  // Usage
  function clientCode(abstractClass) {
    abstractClass.templateMethod();
  }
  
  console.log("Same client code can work with different subclasses:");
  clientCode(new ConcreteClass1());
  
  console.log("\nSame client code can work with different subclasses:");
  clientCode(new ConcreteClass2());
  

File: design_patterns_101/Behavioral Patterns/Template Method Pattern/templateMethodPattern.py
from abc import ABC, abstractmethod

# Abstract Class
class AbstractClass(ABC):
    def template_method(self) -> None:
        self.base_operation1()
        self.required_operation1()
        self.base_operation2()
        self.hook1()
        self.required_operation2()
        self.base_operation3()
        self.hook2()

    def base_operation1(self) -> None:
        print("AbstractClass says: I am doing the bulk of the work")

    def base_operation2(self) -> None:
        print("AbstractClass says: But I let subclasses override some operations")

    def base_operation3(self) -> None:
        print("AbstractClass says: But I am doing the bulk of the work")

    @abstractmethod
    def required_operation1(self) -> None:
        pass

    @abstractmethod
    def required_operation2(self) -> None:
        pass

    def hook1(self) -> None:
        pass

    def hook2(self) -> None:
        pass

# Concrete Class
class ConcreteClass1(AbstractClass):
    def required_operation1(self) -> None:
        print("ConcreteClass1 says: Implemented Operation1")

    def required_operation2(self) -> None:
        print("ConcreteClass1 says: Implemented Operation2")

# Another Concrete Class
class ConcreteClass2(AbstractClass):
    def required_operation1(self) -> None:
        print("ConcreteClass2 says: Implemented Operation1")

    def required_operation2(self) -> None:
        print("ConcreteClass2 says: Implemented Operation2")
        print("ConcreteClass2 says: Override Hook1")

    def hook1(self) -> None:
        print("ConcreteClass2 says: Override Hook2")

# Usage
def client_code(abstract_class: AbstractClass) -> None:
    abstract_class.template_method()

print("Same client code can work with different subclasses:")
client_code(ConcreteClass1())

print("\nSame client code can work with different subclasses:")
client_code(ConcreteClass2())
